source/slang/slang-check-decl.cpp

```diff
@@ -7981,19 +8326,325 @@ namespace Slang
+                }
+                
+                // Part 2
+                // Regular assignment logic
+                paramIndex++;
+                auto paramType = param->getType();
+                auto paramExpr = m_astBuilder->create<VarExpr>();
+                paramExpr->scope = ctor->ownedScope;
+                paramExpr->declRef = param;
+                paramExpr->type = paramType;
+                paramExpr->loc = param->loc;
+
+                // Do not map a variable which is not visible
+                // Do not map a read-only variable for default-init
+                // Do not map compiler generated 'internal processing' variable
+                VarDeclBase* member = members[memberIndex++].getDecl();
+                while (getDeclVisibility(member) < ctorVisibility
+                    || !getTypeForDeclRef(m_astBuilder, member, member->loc).isLeftValue
+                    || member->getName() && member->getName()->text[0] == '$'
+                    )
+                {
+                    // Should note be possible to be out of range unless compiler generated a synth-ctor wrong.
+                    SLANG_ASSERT(memberIndex < members.getCount());
+                    member = members[memberIndex++].getDecl();
+                }
+
+                // We need to ensure member is `no_diff` if it cannot be differentiated, `ctor` modifiers do not matter
+                // in this case since member-wise ctor is always differentiable or "treat as differentiable".
+                if (!isTypeDifferentiable(member->getType()) || member->hasModifier<NoDiffModifier>())
+                {
+                    auto noDiffMod = m_astBuilder->create<NoDiffModifier>();
+                    noDiffMod->loc = param->loc;
+                    addModifier(param, noDiffMod);
+                }
+
+                addCudaHostModifierIfRequired(ctor, member, foundCudaHostModifier);
+
+                MemberExpr* memberExpr = m_astBuilder->create<MemberExpr>();
+                memberExpr->baseExpression = thisExpr;
+                memberExpr->declRef = member->getDefaultDeclRef();
+                memberExpr->scope = ctor->ownedScope;
+                memberExpr->loc = member->loc;
+                memberExpr->name = member->getName();
+                memberExpr->type = DeclRefType::create(getASTBuilder(), memberExpr->declRef);
+
+                auto assign = m_astBuilder->create<AssignExpr>();
+                assign->left = memberExpr;
+                assign->right = paramExpr;
+                assign->loc = paramExpr->loc;
+
+                auto stmt = m_astBuilder->create<ExpressionStmt>();
+                stmt->expression = assign;
+                stmt->loc = assign->loc;
+
+                seqStmtChild->stmts.add(stmt);
+            }
+            seqStmt->stmts.insert(ctorInfo.m_insertOffset++, seqStmtChild);
         }
 
-        if (structDeclInfo.defaultCtor)
+        // Compiler generated ctor may be destroyed
```

Comment:
Why do we need to remove a ctor? Can we just not generate it in the first place?

Reply:
I think this ctor-deletion routine can just be removed completely, you are correct.

We should no longer be creating __init functions which need to be deleted in our bodyVisitor. 

This must hold true since otherwise we would have a mismatch of conformances from the conformance AST-Pass and actual conformances (since deleting a ctor breaks conformances).

Reply:
After reading the code carefully, I found that we cannot avoid removing the ctor based on current change.

This change create a empty default ctor during `SemanticsDeclConformancesVisitor::visitAggTypeDecl`,
and it will fill the body of this default ctor in `SemanticsDeclBodyVisitor::visitAggTypeDecl`.

But there're some situations where it will not fill the body of the ctor, so if that is the case, we will remove the empty default ctor. (Though I still don't quite understand why we have to remove it, removing it will cause lots of errors).

We're actually able to detect those situations during creating the ctor, however the logic to check those situations are part of `SemanticsDeclBodyVisitor::visitAggTypeDecl`, and there is not a very good way to re-use the results of this logic.

So I'd say let's keep it, as cost of removing it could be smaller than "detecting and not create it".